### PR TITLE
[CodeGen][NewPM] Port MachineCSE pass to new pass manager.

### PR DESCRIPTION
**Pull request by cdevadas**

### CODE REVIEW

**Comment by cdevadas** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/106605?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#106605** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/106605?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#106604** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/106604?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @cdevadas and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/include/llvm/CodeGen/MachineCSE.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
+//===- llvm/CodeGen/MachineCSE.h -----------------*- C++ -*-===//
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/MachineCSE.h:1]: Should pad out to 80 lines

File: llvm/lib/CodeGen/MachineCSE.cpp [hunk 2]
```diff
@@ -932,18 +935,55 @@ bool MachineCSE::isProfitableToHoistInto(MachineBasicBlock *CandidateBB,
...
+  MachineBlockFrequencyInfo &MBFI =
+      MFAM.getResult<MachineBlockFrequencyAnalysis>(MF);
+  MachineCSEImpl Impl(&MDT, &MBFI);
+  bool Changed = Impl.run(MF);
+  if (!Changed)
+    return PreservedAnalyses::all();
+
+  auto PA = getMachineFunctionPassPreservedAnalyses();
+  PA.preserve<MachineLoopAnalysis>();
+  PA.preserve<MachineDominatorTreeAnalysis>();
```
Comment (arsenm) [inline llvm/lib/CodeGen/MachineCSE.cpp:973]: These should be redundant with PA.preserveSet<CFGAnalyses>?

Comment (cdevadas) [reply llvm/lib/CodeGen/MachineCSE.cpp:973]: Not sure about that. I have seen similar instances (DT being preserved along with the preserveSet) in other backend passes ported to NPM.

Comment (arsenm) [reply llvm/lib/CodeGen/MachineCSE.cpp:973]: I would say we do not have a representative sample. If you have to explicitly mark these, there's not much point in having the preservesCFG feature

Comment (cdevadas) [reply llvm/lib/CodeGen/MachineCSE.cpp:973]: I guess, we won't be able to figure it out until we have the whole pipeline running in NPM.

**Review by arsenm — APPROVED** [review]

**Comment by cdevadas** [comment]
### Merge activity

* **Sep 4, 8:51 AM EDT**: @cdevadas started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/106605).

**Pull request merged**